### PR TITLE
Rescue IO errors in blocking IO calls caused by non-blocking files.

### PR DESCRIPTION
**Pull request by aardvark179**

### CODE REVIEW

File: src/main/ruby/core/io.rb [hunk 1]
```diff
@@ -158,7 +158,7 @@ def fill(io)
 
       begin
         buffer = Truffle::POSIX.read_string(io.descriptor, count)
-      rescue Errno::EAGAIN
+      rescue Errno::EAGAIN, IO::EAGAINWaitReadable
```
Comment (eregon) [inline src/main/ruby/core/io.rb]: This should just be `IO::WaitReadable` which is the common ancestor of all these kind of "try again" errnos.

File: src/main/ruby/core/posix.rb [hunk 2]
```diff
@@ -292,11 +296,9 @@ def self.write_string(fd, string, nonblock=false)
...
       if ret < 0
-        if nonblock
-          errno = Errno.errno
-          if errno == Errno::EAGAIN::Errno || errno == Errno::EWOULDBLOCK::Errno
-            raise IO::EAGAINWaitWritable
-          end
+        errno = Errno.errno
+        if errno == Errno::EAGAIN::Errno || errno == Errno::EWOULDBLOCK::Errno
+          raise IO::EAGAINWaitWritable
         end
```
Comment (eregon) [inline src/main/ruby/core/posix.rb]: This is problematic as mentioned in the comment below.
If we get EAGAIN after writing a few bytes, the caller cannot know how much we wrote, and we will end up writing the same data twice or more times.

It seems a nice way to handle this is to move the `select/retry` logic here if blocking==true.

File: src/main/ruby/core/io.rb [hunk 3]
```diff
@@ -2631,7 +2631,7 @@ def write(data)
     end
 
     if @sync
-      Truffle::POSIX.write_string @descriptor, data
+      Truffle::POSIX.write_string_with_retry self, data
```
Comment (eregon) [inline src/main/ruby/core/io.rb]: Shouldn't IO#write retry?

File: src/main/ruby/core/posix.rb [hunk 4]
```diff
@@ -272,41 +272,97 @@ def self.with_array_of_strings_pointer(strings)
     end
   end
 
+  def self.read_string_with_retry(io, count)
+    fd = io.descriptor
+    errno = 1
+    until errno == 0
+      (res, errno) = read_string(fd, count)
```
Comment (eregon) [inline src/main/ruby/core/posix.rb]: We don't use this style anywhere else. I'd rather stick to without the parens.

File: src/main/ruby/core/posix.rb [hunk 5]
```diff
@@ -272,41 +272,97 @@ def self.with_array_of_strings_pointer(strings)
...
+        IO.select([], [io])
+      else
+        Errno.handle
+      end
+    end
+  end
+
+  def self.write_string_no_retry(io, string)
+    fd = io.descriptor
+    (written, errno) = write_string(fd, string)
```
Comment (eregon) [inline src/main/ruby/core/posix.rb]: This will actually retry to call `write()` it seems.

File: src/main/ruby/core/io.rb [hunk 6]
```diff
@@ -2219,7 +2214,12 @@ def readpartial(size, buffer=nil)
         return @ibuffer.shift(size)
       end
 
-      sysread(size)
+      begin
+        return sysread(size)
+      rescue Errno::EAGAIN, IO::WaitReadable
+        IO::select([self])
+        retry
+      end
```
Comment (eregon) [inline src/main/ruby/core/io.rb]: Why is this one not using a helper?

File: src/main/ruby/core/io.rb (hunk 3 above)
Comment (aardvark179) [reply src/main/ruby/core/io.rb]: Yes, which is why it calls `write_string_with_retry` and not `write_string_no_retry`.

**Comment by eregon** [comment]
@aardvark179 I added some commits to try the simplify the logic, could you review them?

**Pull request merged**